File: app/src/main/java/com/jeeps/gamecollector/AddPlatformActivity.java
package com.jeeps.gamecollector;

import android.content.Intent;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.RadioButton;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;
import com.jeeps.gamecollector.model.Platform;
import com.jeeps.gamecollector.utils.Colors;
import com.jeeps.gamecollector.utils.FileUtils;
import com.squareup.picasso.Picasso;

import java.io.IOException;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class AddPlatformActivity extends AppCompatActivity {

    public static final String PLATFORM = "PLATFORM";
    public static final String COVER_FILE = "COVER_FILE";

    @BindView(R.id.add_platform_layout) ConstraintLayout rootLayout;
    @BindView(R.id.platform_cover) ImageView platformCover;
    @BindView(R.id.platform_name_edit) EditText platformNameInput;

    private Platform platform;
    private Uri currImageURI;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_platform);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ButterKnife.bind(this);

        platform = new Platform();
        platform.setColor(Colors.NORMIE_WHITE.getColor());

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(view -> createAndReturnPlatform());
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == RESULT_OK) {
            if (requestCode == 1) {
                currImageURI = data.getData();
                // Load local image for displaying purposes
                Picasso.with(this).load(currImageURI).into(platformCover);
                platformCover.setAlpha(1f);
                platformCover.setBackgroundColor(Color.parseColor("#99cccccc"));
            }
        }
    }

    private void createAndReturnPlatform() {
        // Get platform details
        String name = platformNameInput.getText().toString();
        if (name.isEmpty()) {
            dataCheckMessage("Please enter a platform name");
            return;
        }
        platform.setName(name);
        platform.setImageUri(currImageURI.toString());
        try {
            // Return platform details
            Intent data = new Intent();
            data.putExtra(PLATFORM, platform);
            data.putExtra(COVER_FILE, FileUtils.compressImage(this, "temp.png", currImageURI));
            setResult(MainLibraryActivity.ADD_PLATFORM_RESULT, data);
            finish();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void onColorPickerClicked(View view) {
        boolean checked = ((RadioButton) view).isChecked();
        switch (view.getId()) {
            case R.id.color_switch_normiewhite:
                if (checked) platform.setColor(Colors.NORMIE_WHITE.getColor());
                break;
            case R.id.color_switchred:
                if (checked) platform.setColor(Colors.SWITCH_RED.getColor());
                break;
            case R.id.color_xboxgreen:
                if (checked) platform.setColor(Colors.XBOX_GREEN.getColor());
                break;
            case R.id.color_playstationblue:
                if (checked) platform.setColor(Colors.PLAYSTATION_BLUE.getColor());
                break;
        }
    }

    private void dataCheckMessage(String message) {
        Snackbar.make(rootLayout, message, Snackbar.LENGTH_SHORT).show();
    }

    @OnClick(R.id.platform_cover)
    protected void choosePlatformCover(View v) {
        // To open up a gallery browser
        Intent intent1 = new Intent();
        intent1.setType("image/*");
        intent1.setAction(Intent.ACTION_GET_CONTENT);
        startActivityForResult(Intent.createChooser(intent1, "Select Picture"),1);
    }
}

File: app/src/main/java/com/jeeps/gamecollector/remaster/ui/gamePlatorms/AddPlatformActivity.kt
package com.jeeps.gamecollector.remaster.ui.gamePlatorms

import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.RadioButton
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.core.widget.doOnTextChanged
import com.jeeps.gamecollector.R
import com.jeeps.gamecollector.databinding.ActivityAddPlatformBinding
import com.jeeps.gamecollector.databinding.ContentAddPlatformBinding
import com.jeeps.gamecollector.model.Platform
import com.jeeps.gamecollector.remaster.ui.base.BaseActivity
import com.jeeps.gamecollector.remaster.utils.extensions.compressImage
import com.jeeps.gamecollector.remaster.utils.extensions.showToast
import com.jeeps.gamecollector.remaster.utils.extensions.viewBinding
import com.jeeps.gamecollector.utils.PlatformColors
import com.squareup.picasso.Picasso
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.ExperimentalCoroutinesApi

@ExperimentalCoroutinesApi
@AndroidEntryPoint
class AddPlatformActivity : BaseActivity() {

    private val binding by viewBinding(ActivityAddPlatformBinding::inflate)
    private lateinit var content: ContentAddPlatformBinding

    private val viewModel: AddPlatformViewModel by viewModels()

    private val registerForOpenDocument = registerForActivityResult(ActivityResultContracts.OpenDocument()) {
        handleImageSelected(it)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
        setSupportActionBar(binding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        content = binding.content

        // Get platform if being edited
        val platform = intent?.getSerializableExtra(EDITED_PLATFORM) as? Platform
        platform?.let {
            viewModel.setPlatform(it)
            viewModel.isEdit = true
        }

        bindFab()
        bindPlatform()
        bindFormElements()
        bindLoading()

        handleErrorMessages()
        handleServiceEvents()
    }

    private fun handleErrorMessages() {
        viewModel.errorMessage.observe(this) {
            showToast(it)
        }
    }

    private fun bindFab() {
        binding.fab.setOnClickListener {
            viewModel.savePlatform()
        }
    }

    private fun bindPlatform() {
        viewModel.platform.observe(this) { platform ->
            platform.imageUri?.let { uri ->
                if (uri.isNotEmpty()) {
                    Picasso.get()
                        .load(uri)
                        .into(content.platformCover)
                }
            }
            content.platformCover.setBackgroundColor(Color.parseColor("#99cccccc"))
            content.platformNameEdit.setText(platform.name)
            // TODO: Is there a better way to do this?
            PlatformColors.values()
                .find { it.color == platform.color }
                ?.let { content.platformColorRadioGroup.check(it.colorId) }
        }
    }

    private fun bindFormElements() {
        content.platformCover.setOnClickListener {
            registerForOpenDocument.launch(arrayOf("image/*"))
        }

        content.platformNameEdit.doOnTextChanged { text, _, _, _ ->
            viewModel.setPlatformName(text.toString())
        }
    }

    private fun bindLoading() {
        viewModel.isLoading.observe(this) { isLoading ->
            toggleProgressbar(isLoading)
        }
    }

    fun onColorPickerClicked(view: View) {
        if (view !is RadioButton) return
        val checked = view.isChecked

        when (view.id) {
            R.id.color_switch_normiewhite ->
                if (checked) viewModel.setPlatformColor(PlatformColors.NORMIE_WHITE.color)
            R.id.color_switchred ->
                if (checked) viewModel.setPlatformColor(PlatformColors.SWITCH_RED.color)
            R.id.color_xboxgreen ->
                if (checked) viewModel.setPlatformColor(PlatformColors.XBOX_GREEN.color)
            R.id.color_playstationblue ->
                if (checked) viewModel.setPlatformColor(PlatformColors.PLAYSTATION_BLUE.color)
        }
    }

    private fun handleImageSelected(uri: Uri?) {
        if (uri != null) {
            Picasso.get().load(uri).into(content.platformCover)
            content.platformCover.setBackgroundColor(Color.parseColor("#99cccccc"))
            viewModel.imageHasBeenEdited()
            viewModel.currentImageUri = uri
        } else {
            showToast("There was a problem with the image you selected")
        }
    }

    private fun toggleProgressbar(isLoading: Boolean) {
        if (isLoading) {
            binding.addPlatformProgressbar.visibility = View.VISIBLE
            binding.fab.visibility = View.INVISIBLE
        } else {
            binding.addPlatformProgressbar.visibility = View.INVISIBLE
            binding.fab.visibility = View.VISIBLE
        }
    }

    private fun handleServiceEvents() {
        viewModel.isPlatformSaved.observe(this) { event ->
            event.getContentIfNotHandled()?.let { isPlatformSaved ->
                if (isPlatformSaved && viewModel.isImageEdited) {
                    val compressedImage =
                        viewModel.currentImageUri?.let { compressImage("temp.png", it) }
                    viewModel.uploadImageCover(compressedImage)
                } else {
                    viewModel.skipImageUpload()
                }
            }
        }

        viewModel.isImageUploaded.observe(this) { event ->
            event.getContentIfNotHandled()?.let { isImageUploaded ->
                if (isImageUploaded) {
                    finish()
                }
            }
        }
    }

    companion object {
        const val EDITED_PLATFORM = "EDITED PLATFORM"
        const val EDITED_PLATFORM_POSITION = "EDITED PLATFORM POSITION"
        const val PLATFORM = "PLATFORM"
    }
}